lib/checktype.cpp

```diff
@@ -432,9 +432,33 @@ void CheckType::checkFloatToIntegerOverflow()
 {
     logChecker("CheckType::checkFloatToIntegerOverflow");
 
+    std::vector<const Token *> skippedTernaries;
+
     for (const Token *tok = mTokenizer->tokens(); tok; tok = tok->next()) {
         const ValueType *vtint, *vtfloat;
 
+        if (!skippedTernaries.empty() && tok == skippedTernaries.back()) {
+            skippedTernaries.pop_back();
+            while (tok && !Token::Match(tok, ";|)|}")) {
```

Comment:
this kind of loop is not elegant. It's error prone. you would skip too many tokens here: `..?x:y, 123...`. I would prefer that you use the AST instead.

I suggest that you create a utility function `isUnreachableOperand(token)` or something like that. And when a potential bug is found in checkFloatToInteger you can call that utility function before writing the warning.

I envision that the `isUnreachableOperand(token)` would walk up the AST and see if there was a parent ternary operator.. It could also lookout for `&&` and `||`.

This `isUnreachableOperand` could be useful in other checkers also.

Reply:
Sounds good. How do you feel about matching with boolean literals? It's maybe a bit hacky, but the value for true isn't set to known in this particular case.

Reply:
> It's maybe a bit hacky, but the value for true isn't set to known in this particular case.

Not sure how that can happen. Or is it only in the test?
![image](https://github.com/user-attachments/assets/58b10feb-214f-43cc-ae57-12a6f2e794b2)

Reply:
I think it was to do with the fact that it's inside a macro expansion. There's a similar thing with template arguments if I'm not mistaken. What did the code look like for your result?

Reply:
I just used one of your test cases.

Reply:
It seems to be dependent on wether or not it's C or C++. I think the distinction is made in vf_number.cpp, I guess it's because `true/false` aren't keywords in C.

Reply:
Actually that does mean matching with true/false isn't correct, I should probably remove that, right?